### ISSUE TITLE
Instant Search SSR fetches data on server and client

### ISSUE THREAD

**Issue by gwardwell** [algolia/vue-instantsearch#763]
**Bug 🐞**

### What is the current behavior?

TL;DR: Algolia SSR setup results in duplicate queries, one on the server side and another on the client side.

When Algolia data is fetched on the server side, I write it to `window.__ALGOLIA_STATE__`:

```js
context.algoliaState = instantsearch.getState();
```

Then, on the client side, I hydrate from that in-page state:

```js
instantsearch.hydrate(window.__ALGOLIA_STATE__);
```

The data is properly fetched on the server side, and I see a populated `window.__ALGOLIA_STATE__` variable in the server response. But when the application boots up on the client side, I see an Algolia query execute. This doesn't seem like the expected behavior.

### Make a sandbox with the current behavior

You can see this behavior when using the [Algolia Vue InstantSearch server-side-rendering demo project](https://github.com/algolia/doc-code-samples/tree/master/Vue%20InstantSearch/server-side-rendering)

After seeing this behavior in my local environment, I pulled that project down to verify the behavior.

I used the current version of Vue, SSR, etc., but upgraded algoliasearch and vue-instantsearch to their current versions.

### What is the expected behavior?

There would be no client-side query on a fresh page load because the data would be hydrated from the server side.

### Does this happen only in specific situations?

No.

### What is the proposed solution?

When hydrating from server side, do not run a duplicate query.

### What is the version you are using?

*  algoliasearch: 4.0.3
*  vue-instantsearch: 2.7.0

> Always try the latest one before opening an issue.

**Comment by Haroenv** [comment]
Sorry @gwardwell, this is a known limitation for now. The workaround you have on the client is explicitly setting the cache value on the client. 

In Vue InstantSearch we do not have access to it like we do on React InstantSearch, where it's implemented like this:

https://github.com/algolia/react-instantsearch/blob/2e8214fd19d5c618cb3013790193747329899a9f/packages/react-instantsearch-core/src/core/createInstantSearchManager.js#L319-L468

**Comment by gwardwell** [comment]
@Haroenv Thanks for the response and explanation.

What's preventing Vue InstantSearch from looking at `window. __ALGOLIA_STATE__` in a similar way too how you're handling hydration in React InstantSearch? Or utilizing the Vue plugin architecture to control state server side caching and client side hydration like @Akryum does with [Vue CLI Plugin Apollo](https://github.com/Akryum/vue-cli-plugin-apollo)?

My big concern is that we'll get a server/client DOM mismatch from loading the data twice. Is the client side hydration accomplishing anything? Or should `instantsearch.hydrate(window.__ALGOLIA_STATE__);` be removed from my code?

**Comment by Haroenv** [comment]
Nothing is preventing us from doing that here, except that timeline wise, Vue InstantSearch server side rendering was made before the solution was found in React InstantSearch, and hasn't been ported to this flavour yet.

It's not a question of impossibility, more a question of time line

**Comment by frostydustpile** [comment]
I believe this same behavior is exhibited in the latest and improved version 3 of Vue Instant Search. So I'm guessing this feature, though understood, didn't make it in the priority list for the V2 to V3 update.

You can see it happening in the Algolia demo at:
https://00j1788yj0.sse.codesandbox.io/search

Or by following the SSR tutorial which produces the same result:
https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/vue/

Is there a way to manually hydrate the cache until a fix from Algolia comes along?

The duplicate (client side) request causes more data over the wire and artificially inflates the number of Algolia queries the website makes. I suppose since Algolia pricing is based on query count, this isn't such a bad thing for Algolia! 😉

**Comment by francoischalifour** [comment]
> Is there a way to manually hydrate the cache until a fix from Algolia comes along?

Assuming you're using `algoliasearch` v4, you can read more on our [cache section](https://www.algolia.com/doc/api-client/advanced/configure-the-client/javascript/?language=javascript#cache-on-requests-and-responses).

> The duplicate (client side) request causes more data over the wire and artificially inflates the number of Algolia queries the website makes. I suppose since Algolia pricing is based on query count, this isn't such a bad thing for Algolia! 😉

That's a heavy statement. That's definitely not our team's intention; our libraries aim at minimizing the cost while providing complete search features. Sometimes, these optimizations come after other features that need to be shipped faster.

It wasn't prioritized enough yet because we want to solve this duplicate request issue at a generic level by providing a SSR API shared among all InstantSearch flavors (InstantSearch.js, Vue InstantSearch, React InstantSearch and Angular InstantSearch). In the meantime though, we should write a guide available on our documentation to help users avoid this second request.

**Comment by frostydustpile** [comment]
Wasn't trying to be too "heavy" - thus the winky. I know it's not your intent to boost requests - just an unfortunate side effect. I'll reread the cache doc. I'm just getting started with Algolia, so it wasn't entirely clear to me the first time (or two) through.

**Comment by Haroenv** [comment]
This is no longer the case since https://github.com/algolia/vue-instantsearch/pull/1090
